framework/resource/resource.go

```diff
@@ -149,6 +152,34 @@ func (r *Resource) Destroy(args ...interface{}) error {
 	return result.Err()
 }
 
+// DeclaredResource converts a resource to a DeclaredResource protobuf, which
+// can be used in a component.DeclaredResourcesResp.
+func (r *Resource) DeclaredResource() (*pb.DeclaredResource, error) {
+	stateJson, err := json.Marshal(r.State())
+	if err != nil {
+		return nil, fmt.Errorf("state for resource is not serializable to json: %w", err)
+	}
+
+	stateProtoAny, err := component.ProtoAny(r.State())
+	if err != nil {
+		return nil, fmt.Errorf("state for resource is not serializable to json: %w", err)
```

Comment:
nit: I feel this error message is misleading; we verified that the state is serializable to json at the start of this method (~line 158-161).